crates/ra_hir/src/ty/method_resolution.rs

```diff
@@ -131,83 +131,96 @@ impl Ty {
         // (and we need to do autoderefs for trait method calls as well)
-        let inherent_method = self.clone().iterate_methods(db, |ty, f| {
+        self.clone().iterate_method_candidates(db, resolver, |ty, f| {
             let sig = f.signature(db);
             if sig.name() == name && sig.has_self_param() {
                 Some((ty.clone(), f))
             } else {
                 None
             }
-        });
-        inherent_method.or_else(|| self.lookup_trait_method(db, name, resolver))
+        })
     }
 
-    fn lookup_trait_method(
+    // This would be nicer if it just returned an iterator, but that runs into
+    // lifetime problems, because we need to borrow temp `CrateImplBlocks`.
+    pub(crate) fn iterate_method_candidates<T>(
         self,
         db: &impl HirDatabase,
-        name: &Name,
         resolver: &Resolver,
-    ) -> Option<(Ty, Function)> {
-        let mut candidates = Vec::new();
+        mut callback: impl FnMut(&Ty, Function) -> Option<T>,
+    ) -> Option<T> {
+        // For method calls, rust first does any number of autoderef, and then one
+        // autoref (i.e. when the method takes &self or &mut self). We just ignore
+        // the autoref currently -- when we find a method matching the given name,
+        // we assume it fits.
+
+        // Also note that when we've got a receiver like &S, even if the method we
+        // find in the end takes &self, we still do the autoderef step (just as
+        // rustc does an autoderef and then autoref again).
+
+        for derefed_ty in self.autoderef(db) {
+            if let Some(result) = derefed_ty.iterate_inherent_methods(db, &mut callback) {
+                return Some(result);
+            }
+            if let Some(result) =
+                derefed_ty.iterate_trait_method_candidates(db, resolver, &mut callback)
+            {
+                return Some(result);
+            }
+        }
+        None
+    }
+
+    fn iterate_trait_method_candidates<T>(
+        &self,
+        db: &impl HirDatabase,
+        resolver: &Resolver,
+        mut callback: impl FnMut(&Ty, Function) -> Option<T>,
+    ) -> Option<T> {
         for t in resolver.traits_in_scope() {
             let data = t.trait_data(db);
             for item in data.items() {
                 match item {
                     &TraitItem::Function(m) => {
-                        let sig = m.signature(db);
-                        if sig.name() == name && sig.has_self_param() {
-                            candidates.push((t, m));
+                        let trait_ref = TraitRef {
+                            trait_: t,
+                            substs: fresh_substs_for_trait(db, t, self.clone()),
+                        };
+                        let (trait_ref, _) = super::traits::canonicalize(trait_ref);
+                        if db.implements(trait_ref).is_none() {
+                            continue;
+                        }
```

Comment:
Can we pull this filtering bit before we interate methods?

Reply:
Also, I was just thinking about rustc's logic of "let's filter by method name first", which has an obvious implication for completion... Perhaps, if we add an `hint: Option<Name>` parameter to iterate functions, we'll get the API which would work nicely for both resolve and completion?

Reply:
Ok, I changed it so we're as lazy as possible ;) Though I think this will still need to change a bit anyway to get the correct priority order of methods...